src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java

```diff
@@ -481,11 +481,7 @@ public Filter termsFilter(List values, @Nullable QueryParseContext context) {
             // bit set
             return termFilter(values.get(0), context);
         default:
-            BytesRef[] bytesRefs = new BytesRef[values.size()];
-            for (int i = 0; i < bytesRefs.length; i++) {
-                bytesRefs[i] = indexedValueForSearch(values.get(i));
-            }
-            return new TermsFilter(names.indexName(), bytesRefs);
+            return new DeferredTermsFilter(names.indexName(), values, this);
```

Comment:
As mentioned in the PR description this is where step (2) happens.  It turns out the `TermsFilter` constructor is quite expensive.  `DeferredTermsFilter` is a wrapper around `TermsFilter` that doesn't instantiate `TermsFilter` until it turns out that we actually need to use it (when there's a cache miss in step (3)).